### DISCUSSION TITLE
Keyboard/Mouse IO

### DISCUSSION THREAD

**Discussion by ryanopily** [UNC-Project-Open-AAC/OS-DPI#104] [Ideas]
I've been experimenting with logging capabilities on my own fork of OS-DPI, and I think it's necessary to be able to log keyboard and mouse events. Perhaps fields (#mouseX, #mouseY, #keyPressed) accessible from the actions tab (origin: *, event: keydown, etc.) could be the way to go. Is this a good idea?

I have gesture recognition software that simulates keystrokes whenever it detects a gesture. For example, smiling may be bound the the 'F1' key. I would log them whenever this key is pressed; also, logging keystrokes and mouseclicks may be beneficial for data analysis purposes. I think the best way of doing this is through an action. 

Please let me know what you think.

**Comment by gbishop** [comment]
Excellent! Just the kind of thinking we need. 

In the [Access](https://github.com/gbishop/OS-DPI/tree/access) branch of my repo, check out [components/access/method/](https://github.com/gbishop/OS-DPI/tree/access/components/access/method). We've got Handlers for various kinds of low-level events. The Handlers map these browser level events into the events we hand to the Actions tab.

As an aside, I'm thinking the "Event" column of the Actions tab might go away. The type of the event will just be another virtual field coming with the rest of the info.

There is a [keyHandler](https://github.com/gbishop/OS-DPI/tree/access/components/access/method) down there as well as pointerHandler and timerHandler. The various handlers gather the event and any data associated with it and hand it off to a response defined in responses.js. 

To the degree that it works, it works just like you suggest with a variety of virtual fields tagged onto the event that gets handed up. 

I see several possible approaches.

1. Do everything in the current Actions panel. If you want to log F1, you just include a rule. Straightforward but repetitive.
2. Allow logging in the handlers responses. There you might log all or any subset of events so the upper level doesn't have to fool with doing it over and over.
3. A hybrid of those two so you can log wherever you wish.
4. Allow "hooking" of the browser events to log them even before the Handlers get to look at them. You'd get a ton of stuff logged but if you're interested in how many debounced events happened that would be the way you could know.

My hope is that most users don't have to fool with the Access tab except to choose the preprogrammed Method and define their Pattern but I think we can provide enough power there for people to do innovative things.

What do you think? It will help me a **TON** to have you thinking about how you would use this stuff.

**Reply by gbishop** [reply]
One more thing. The code in the Handlers is using [rxjs](https://rxjs.dev/) to process what they call streams. It took me a while to wrap my head around it. If you'd like to meet to talk about what is happening down there let me know. Their documentation is really good.

**Reply by ryanopily** [reply]
I was able to figure out how to log keystrokes using the access tab, and it seems like the perfect solution to my problem.

I think that access is great not only as an accessibility feature, but also as a programming tool; I've already said that I want to use it for logging, but I think that it could be useful for other things in the future too, though I don't have any other ideas on what it could be used for just yet. 

Awesome work, and thanks for the help!

**Comment by gbishop** [comment]
Thanks Ryan. That is very encouraging! I've been the only one thinking
about this stuff and I wondered if it was going to work for anyone else.

I hope to merge this into main this week.
